Add automation to support building and pushing image to Docker Hub automatically

Muscaw opened the pull request:
This PR provides the ability to build and push the `latest` docker image when some conditions are met.

The pipeline executes on the following triggers:
- a PR is created against the repo (by default, PRs from forks are not executed for security reasons)
- Whenever the master branch is updated
- On tag creation in the format x.y.z (semver)

The docker image is built in all of these scenarios, but **only** when a tag occurs in the format described above is the image pushed to Docker Hub.

To work as expected, the secrets from DockerHub must be set on the repository. This guide supports the creation of the secrets: https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions?tool=webui#creating-secrets-for-a-repository

The names must be exactly as follows:
- DOCKERHUB_USERNAME: will be the username of the owner of the docker hub account where the images are pushed
- DOCKERHUB_TOKEN: will be the token created to push the images. A new token can be created using this link (when logged in to Docker hub) https://hub.docker.com/settings/security?generateToken=true

The permissions required for DOCKERHUB_TOKEN is `Read & Write`

wernight commented:
I agree with the idea, however I don't see how the current workflow achieves this. Especially for an encrypted backup tool, I want to make sure it's as safe as possible.

If it's too hard to ensure that the logic is safe, I'd prefer to only run CI on the master branch; at the cost of CI giving less info during a PR.

Muscaw commented:
By default, workflows are not executed against forks if not approved specifically. See https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks

If you prefer, I can simply remove the PR event from the workflow and only master branch and tags would trigger the workflow

wernight commented:
Ah, that's a decent security feature indeed.

Would you mind replying on both inline comments.

Muscaw commented:
@wernight I'm sorry, but I don't see inline comments on the PR

wernight commented:
A) Is this required?

```
    defaults:
      run:
        shell: bash
```

B) Could you add a comment before the line below to explain why "refs/tags":

```
push: ${{ startsWith(github.ref, 'refs/tags') }}
```

Muscaw commented:
@wernight I removed the defaults shell. For non-windows hosts, it defaults to bash

I also added a comment before the push line.
Let me know if you see anything else

wernight commented:
Thank you.
